File: problems/Counting Bits/countBits.cs
public class Solution {
    public int[] CountBits(int num) {
        var result = new int[1 + num];

        for (var i = 1; num >= i; ++i) {
            result[i] = result[i >> 1] + (i & 1);
        }

        return result;
    }
}

File: problems/Counting Bits/countBits.java
class Solution {
    public int[] countBits(int n) {
        int[] ans = new int[n + 1];
        int x = 0;
        int b = 1;

        // [0, b) is calculated
        while (b <= n) {
            // generate [b, 2b) or [b, n) from [0, b)
            while (x < b && x + b <= n) {
                ans[x + b] = ans[x] + 1;
                ++x;
            }
            x = 0; // reset x
            b <<= 1; // b = 2b
        }

        return ans;
    }
}
